nixos/modules/security/please.nix

```diff
@@ -0,0 +1,106 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.security.please;
+  ini = pkgs.formats.ini { };
+in
+{
+  options.security.please = {
+    enable = mkEnableOption (mdDoc ''
+      please, a Sudo clone which allows a users to execute a command or edit a
+      file as another user
+    '');
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.please;
+      defaultText = literalExpression "pkgs.please";
+      description = mdDoc ''
+        Which package to use for {command}`please`;
+      '';
+    };
+
+    wheelNeedsPassword = mkOption {
+      type = types.bool;
+      default = true;
+      description = lib.mdDoc ''
+        Whether users of the `wheel` group must provide a password to run
+        commands or edit files with {command}`please` and
+        {command}`pleaseedit` respectively.
+      '';
+    };
+
+    settings = mkOption {
+      type = ini.type;
+      default = { };
+      example = {
+        jim_become_root = {
+          name = "jim";
+          target = "root";
+          rule = ".*";
+          require_pass = false;
+        };
+        jim_hosts = {
+          name = "jim";
+          type = "edit";
+          target = "root";
+          rule = "/etc/hosts";
+          editmode = 644;
+          require_pass = false;
+        };
+      };
+      description = mdDoc ''
+        Please configuration. Refer to
+        <https://github.com/edneville/please/blob/master/please.ini.md> for
+        details.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    security.wrappers =
+      let
+        owner = "root";
+        group = "root";
+        setuid = true;
+      in
+      {
+        please = {
+          source = "${cfg.package}/bin/please";
+          inherit owner group setuid;
+        };
+        pleaseedit = {
+          source = "${cfg.package}/bin/pleaseedit";
+          inherit owner group setuid;
+        };
+      };
+
+    security.please.settings = rec {
+      wheel_run_as_root = {
+        type = "run";
+        name = "wheel";
+        group = true;
+        target = "root";
+        rule = ".*";
+        require_pass = cfg.wheelNeedsPassword;
+      };
+      wheel_edit_as_root = wheel_run_as_root // { type = "edit"; };
+      wheel_list_as_root = wheel_run_as_root // { type = "list"; };
+    };
+
+    environment = {
+      systemPackages = [ cfg.package ];
+
+      etc."please.ini".source = ini.generate "please.ini" cfg.settings;
```

Comment:
Oh, we'll probably also want a rule that allows root to do anything using please. I don't know if please does this implicitly, but if not, it's something that should be added.

I'd argue it should go here, like `({ root_run = { needs_password = false; ... }; root_edit = { ... }; root_list = { ... }; } // cfg.settings)` so that users can't overwrite it (though the message when providing a same-named rule won't be that pretty) -- doas does something similar by just hardcoding `permit nopass keepenv root` in the generated file.

Reply:
Now `root` can do anything by default. I couldn't make mkForce or something similar to work for the INI generator, so I just appended this configuration to the exposed option via //: this will ignore attributes with the same name on generation and won't error out (which I don't think is ideal, but I can't find other solutions to this).